{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in sv2 merged table\n",
    "# should have columns ['chrom', 'svtype', 'pos', 'end', 'svlength']\n",
    "# can have more columns\n",
    "svs = pd.read_csv('output/merged.sv2.breakpoints_resolved.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0,22,24,36,37,46,47,61,62,72,73,82,83,96,97,106,107,116,117,130,131,140,141,150,151,164,165,174,175,184,185,198,199,208,209,218,219,232,233) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# downloaded from https://gnomad.broadinstitute.org/downloads#v2-structural-variants\n",
    "gnomad = pd.read_csv('gnomad_v2.1_sv.nonneuro.sites.bed.gz', sep='\\t')\n",
    "\n",
    "# create gnomad column for chromosome with chr\n",
    "gnomad['chrom'] = gnomad['#chrom'].apply(lambda s: 'chr{}'.format(s))\n",
    "\n",
    "# filter for only dups and dels\n",
    "gnomad = gnomad[gnomad.svtype.isin(['DEL', 'DUP'])].copy()\n",
    "\n",
    "# filter for pass only\n",
    "gnomad = gnomad[gnomad.FILTER == 'PASS'].copy()\n",
    "\n",
    "# set gnomad frequency to type float\n",
    "gnomad['AF'] = gnomad.AF.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop over every chrom and svtype\n",
    "chromosomes = list(svs.chrom.unique())\n",
    "svtypes = ['DUP', 'DEL']\n",
    "svs['gnomad_freq_non_neuro'] = '.'\n",
    "\n",
    "for chrom in chromosomes:\n",
    "    for svtype in svtypes:\n",
    "        # get all strctural variants in chromosome and svtype\n",
    "        sub_svs = svs[(svs.chrom == chrom) & (svs.svtype == svtype)].copy()\n",
    "        sub_gnomad = gnomad[(gnomad['chrom'] == chrom) & (gnomad.svtype == svtype)].copy()\n",
    "        \n",
    "        # reset index in sub_gnomad\n",
    "        sub_gnomad = sub_gnomad.reset_index(drop=True)\n",
    "        \n",
    "        # create numpy structures for end, start, and svlength\n",
    "        # (numpy is faster than pandas)\n",
    "        chrom_ends = sub_gnomad['end'].to_numpy()\n",
    "        chrom_starts = sub_gnomad['start'].to_numpy()\n",
    "        chrom_svlengths = chrom_ends - chrom_starts\n",
    "\n",
    "        # iterate over all structural variants in cohort\n",
    "        # and compare against other structural variants gnomad\n",
    "        for i, row in sub_svs.iterrows():\n",
    "            start = row['start']\n",
    "            end = row['end']\n",
    "            length = row['svlength']\n",
    "\n",
    "            min_end   = np.minimum(chrom_ends, end)\n",
    "            max_start = np.maximum(chrom_starts, start)\n",
    "            max_length = np.maximum(chrom_svlengths, length)\n",
    "\n",
    "            overlap = (min_end - max_start) > .5 * max_length\n",
    "\n",
    "            # in some cases the structural variant overlaps with more than one gnomad SV\n",
    "            # in those cases get the sum of all SV that it overlaps with\n",
    "            allele_frequency = sub_gnomad.loc[overlap]['AF'].sum()\n",
    "            \n",
    "            # save to table\n",
    "            svs.at[i, 'gnomad_freq_non_neuro'] = allele_frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "svs.to_csv('output/merged.sv2.gnomad_non_neuro.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>svtype</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>svlength</th>\n",
       "      <th>genes</th>\n",
       "      <th>gnomad_freq_non_neuro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SG172</td>\n",
       "      <td>DUP</td>\n",
       "      <td>chr1</td>\n",
       "      <td>845001</td>\n",
       "      <td>850200</td>\n",
       "      <td>5200</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0.000497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SG172</td>\n",
       "      <td>DUP</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13178567</td>\n",
       "      <td>13187129</td>\n",
       "      <td>8563</td>\n",
       "      <td>NM_001136561,LOC440563,exon_1/1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SG172</td>\n",
       "      <td>DUP</td>\n",
       "      <td>chr1</td>\n",
       "      <td>31070606</td>\n",
       "      <td>31070669</td>\n",
       "      <td>64</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0.327312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SG172</td>\n",
       "      <td>DUP</td>\n",
       "      <td>chr1</td>\n",
       "      <td>56996032</td>\n",
       "      <td>56996346</td>\n",
       "      <td>315</td>\n",
       "      <td>NM_003713,PPAP2B,intron_1/5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SG172</td>\n",
       "      <td>DUP</td>\n",
       "      <td>chr1</td>\n",
       "      <td>59644931</td>\n",
       "      <td>59670360</td>\n",
       "      <td>25430</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422002</th>\n",
       "      <td>SG445</td>\n",
       "      <td>DEL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>151982765</td>\n",
       "      <td>151982840</td>\n",
       "      <td>76</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0.764914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422003</th>\n",
       "      <td>SG445</td>\n",
       "      <td>DEL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152807507</td>\n",
       "      <td>152807579</td>\n",
       "      <td>73</td>\n",
       "      <td>NM_001001344,ATP2B3,intron_1/19|NM_021949,ATP2...</td>\n",
       "      <td>0.233868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422004</th>\n",
       "      <td>SG445</td>\n",
       "      <td>DEL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154365486</td>\n",
       "      <td>154365579</td>\n",
       "      <td>94</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>0.459274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422005</th>\n",
       "      <td>SG445</td>\n",
       "      <td>DEL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154771601</td>\n",
       "      <td>154785600</td>\n",
       "      <td>14000</td>\n",
       "      <td>NM_018196,TMLHE,exon_1/7|NM_018196,TMLHE,intro...</td>\n",
       "      <td>0.0123515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422006</th>\n",
       "      <td>SG445</td>\n",
       "      <td>DEL</td>\n",
       "      <td>chrX</td>\n",
       "      <td>154790190</td>\n",
       "      <td>154803524</td>\n",
       "      <td>13335</td>\n",
       "      <td>NM_018196,TMLHE,UTR5|NM_001184797,TMLHE,UTR5</td>\n",
       "      <td>0.857326</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422007 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample svtype chrom      start        end  svlength  \\\n",
       "0       SG172    DUP  chr1     845001     850200      5200   \n",
       "1       SG172    DUP  chr1   13178567   13187129      8563   \n",
       "2       SG172    DUP  chr1   31070606   31070669        64   \n",
       "3       SG172    DUP  chr1   56996032   56996346       315   \n",
       "4       SG172    DUP  chr1   59644931   59670360     25430   \n",
       "...       ...    ...   ...        ...        ...       ...   \n",
       "422002  SG445    DEL  chrX  151982765  151982840        76   \n",
       "422003  SG445    DEL  chrX  152807507  152807579        73   \n",
       "422004  SG445    DEL  chrX  154365486  154365579        94   \n",
       "422005  SG445    DEL  chrX  154771601  154785600     14000   \n",
       "422006  SG445    DEL  chrX  154790190  154803524     13335   \n",
       "\n",
       "                                                    genes  \\\n",
       "0                                              intergenic   \n",
       "1                         NM_001136561,LOC440563,exon_1/1   \n",
       "2                                              intergenic   \n",
       "3                             NM_003713,PPAP2B,intron_1/5   \n",
       "4                                              intergenic   \n",
       "...                                                   ...   \n",
       "422002                                         intergenic   \n",
       "422003  NM_001001344,ATP2B3,intron_1/19|NM_021949,ATP2...   \n",
       "422004                                         intergenic   \n",
       "422005  NM_018196,TMLHE,exon_1/7|NM_018196,TMLHE,intro...   \n",
       "422006       NM_018196,TMLHE,UTR5|NM_001184797,TMLHE,UTR5   \n",
       "\n",
       "       gnomad_freq_non_neuro  \n",
       "0                   0.000497  \n",
       "1                          0  \n",
       "2                   0.327312  \n",
       "3                          0  \n",
       "4                          0  \n",
       "...                      ...  \n",
       "422002              0.764914  \n",
       "422003              0.233868  \n",
       "422004              0.459274  \n",
       "422005             0.0123515  \n",
       "422006              0.857326  \n",
       "\n",
       "[422007 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "svs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
